File: libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/internal/AutoPersistedQueryInterceptor.java
package com.apollographql.apollo3.runtime.java.internal;

import com.apollographql.apollo3.api.ApolloRequest;
import com.apollographql.apollo3.api.ApolloResponse;
import com.apollographql.apollo3.api.Mutation;
import com.apollographql.apollo3.api.Operation;
import com.apollographql.apollo3.api.http.HttpMethod;
import com.apollographql.apollo3.exception.ApolloException;
import com.apollographql.apollo3.exception.AutoPersistedQueriesNotSupported;
import com.apollographql.apollo3.runtime.java.ApolloCallback;
import com.apollographql.apollo3.runtime.java.ApolloInterceptor;
import com.apollographql.apollo3.runtime.java.ApolloInterceptorChain;
import com.apollographql.apollo3.api.Error;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class AutoPersistedQueryInterceptor implements ApolloInterceptor {
  private final HttpMethod httpMethodForHashedQueries;
  private final HttpMethod httpMethodForDocumentQueries;

  public AutoPersistedQueryInterceptor(
      HttpMethod httpMethodForHashedQueries,
      HttpMethod httpMethodForDocumentQueries
  ) {
    this.httpMethodForHashedQueries = httpMethodForHashedQueries;
    this.httpMethodForDocumentQueries = httpMethodForDocumentQueries;
  }

  @Override public <D extends Operation.Data> void intercept(@NotNull ApolloRequest<D> request, @NotNull ApolloInterceptorChain chain, @NotNull ApolloCallback<D> callback) {
    Boolean enabled = request.getEnableAutoPersistedQueries();
    if (enabled == null) {
      enabled = true;
    }

    if (!enabled) {
      chain.proceed(request, callback);
    }

    boolean isMutation = request.getOperation() instanceof Mutation;

    ApolloRequest<D> request2 = request.newBuilder()
        .httpMethod(isMutation ? HttpMethod.Post : httpMethodForHashedQueries)
        .sendDocument(false)
        .sendApqExtensions(true)
        .build();


    ApolloCallback<D> callback2 = new ApolloCallback<D>() {
      @Override public void onResponse(@NotNull ApolloResponse<D> response) {
        if (isPersistedQueryNotFound(response.errors)) {
          continueWithDocumentRequest(chain, request, callback);
        } else if (isPersistedQueryNotSupported(response.errors)) {
          callback.onFailure(new AutoPersistedQueriesNotSupported());
        } else {
          // Cache hit
          callback.onResponse(addAutoPersistedQueryInfo(response, true));
        }
      }

      @Override public void onFailure(@NotNull ApolloException e) {
        callback.onFailure(e);
      }
    };

    chain.proceed(request2, callback2);
  }

  private <D extends Operation.Data> void continueWithDocumentRequest(ApolloInterceptorChain chain, ApolloRequest<D> request, ApolloCallback<D> callback) {
    boolean isMutation = request.getOperation() instanceof Mutation;
    ApolloRequest<D> request2 = request.newBuilder()
        .httpMethod(isMutation ? HttpMethod.Post : httpMethodForDocumentQueries)
        .sendDocument(true)
        .sendApqExtensions(true)
        .build();

    ApolloCallback<D> callback2 = new ApolloCallback<D>() {
      @Override public void onResponse(@NotNull ApolloResponse<D> response) {
        callback.onResponse(addAutoPersistedQueryInfo(response, false));
      }

      @Override public void onFailure(@NotNull ApolloException e) {
        callback.onFailure(e);
      }
    };

    chain.proceed(request2, callback2);
  }

  static private <D extends Operation.Data> ApolloResponse<D> addAutoPersistedQueryInfo(ApolloResponse<D> response, boolean hit) {
    return response.newBuilder()
        .addExecutionContext(new AutoPersistedQueryInfo(hit))
        .build();
  }

  static private boolean isPersistedQueryNotFound(List<Error> errors) {
    if (errors == null) {
      return false;
    }
    for (Error error : errors) {
      if (error.getMessage().equals(PROTOCOL_NEGOTIATION_ERROR_QUERY_NOT_FOUND)) {
        return true;
      }
    }
    return false;
  }

  static private boolean isPersistedQueryNotSupported(List<Error> errors) {
    if (errors == null) {
      return false;
    }
    for (Error error : errors) {
      if (error.getMessage().equals(PROTOCOL_NEGOTIATION_ERROR_NOT_SUPPORTED)) {
        return true;
      }
    }
    return false;
  }

  private static final String PROTOCOL_NEGOTIATION_ERROR_QUERY_NOT_FOUND = "PersistedQueryNotFound";
  private static final String PROTOCOL_NEGOTIATION_ERROR_NOT_SUPPORTED = "PersistedQueryNotSupported";
}

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/AutoPersistedQueryInterceptor.kt
package com.apollographql.apollo3.interceptor

import com.apollographql.apollo3.AutoPersistedQueryInfo
import com.apollographql.apollo3.api.ApolloRequest
import com.apollographql.apollo3.api.ApolloResponse
import com.apollographql.apollo3.api.Error
import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.http.HttpMethod
import com.apollographql.apollo3.exception.AutoPersistedQueriesNotSupported
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.single

internal class AutoPersistedQueryInterceptor(
    private val httpMethodForHashedQueries: HttpMethod,
    private val httpMethodForDocumentQueries: HttpMethod,
) : ApolloInterceptor {

  override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
    val enabled = request.enableAutoPersistedQueries ?: true

    if (!enabled) {
      return chain.proceed(request)
    }

    @Suppress("NAME_SHADOWING")
    val request = request.newBuilder()
        .httpMethod(httpMethodForHashedQueries)
        .sendDocument(false)
        .sendApqExtensions(true)
        .build()

    return flow {
      var response = chain.proceed(request).single()

      when {
        isPersistedQueryNotFound(response.errors) -> {
          val retryRequest = request
              .newBuilder()
              .httpMethod(httpMethodForDocumentQueries)
              .sendDocument(true)
              .sendApqExtensions(true)
              .build()

          response = chain.proceed(retryRequest).single()
          emit(response.withAutoPersistedQueryInfo(false))
        }
        isPersistedQueryNotSupported(response.errors) -> {
          throw AutoPersistedQueriesNotSupported()
        }
        else -> {
          emit(response.withAutoPersistedQueryInfo(true))
        }
      }
    }
  }

  private fun <D : Operation.Data> ApolloResponse<D>.withAutoPersistedQueryInfo(hit: Boolean) = newBuilder()
      .addExecutionContext(AutoPersistedQueryInfo(hit))
      .build()

  private fun isPersistedQueryNotFound(errors: List<Error>?): Boolean {
    return errors?.any { it.message.equals(PROTOCOL_NEGOTIATION_ERROR_QUERY_NOT_FOUND, ignoreCase = true) } == true
  }

  private fun isPersistedQueryNotSupported(errors: List<Error>?): Boolean {
    return errors?.any { it.message.equals(PROTOCOL_NEGOTIATION_ERROR_NOT_SUPPORTED, ignoreCase = true) } == true
  }

  companion object {
    private const val PROTOCOL_NEGOTIATION_ERROR_QUERY_NOT_FOUND = "PersistedQueryNotFound"
    private const val PROTOCOL_NEGOTIATION_ERROR_NOT_SUPPORTED = "PersistedQueryNotSupported"
  }
}